src/bulk-load.js

```diff
@@ -115,6 +127,33 @@ module.exports = class BulkLoad extends EventEmitter {
     }
   }
 
+  /**
+   * @param {{checkConstraints : boolean , fireTriggers : boolean , keepNulls : boolean, tableLock: boolean}} bulkLoadOptions
+   * @descriptions
+   *  checkConstraints - honors constraints during bulk load, it is disabled by default.
+   *
+   *  fireTriggers -  honors insert triggers during bulk load, it is disabled by default.
+   *
+   *  keepNulls - honors null value passed, ignores the default values set on table.
+   *
+   *  tableLock - places a bulk update(BU) lock on table while performing bulk load. Uses row locks by default
+   */
+  setOptions(bulkLoadOptions) {
+    if (bulkLoadOptions) {
```

Comment:
Iterating over object properties using `for ... in` is problematic in JavaScript, mostly because _any_ object can be extended with random properties via extensions on `Object.prototype`. Also, I believe it's quite unexpected that an error gets thrown on "unknown" properties.

Another issue here is that `bulkOptions` is modified during the iteration of `bulkLoadOptions`, which can lead the bulk load options to be left in an unexpected state when one of the properties in `bulkLoadOptions` has an invalid value.

What do you think about changing this code to follow precedent set in the Node.js core codebase? E.g. if we take this snippet here https://github.com/nodejs/node/blob/556ebab30e171c2941deb529a42ad47301972af5/lib/child_process.js#L378-L489 as an inspiration, the `setOptions` method could look something like this:

```js
setOptions(options) {
  if (typeof options !== 'object') {
    throw new TypeError('"options" argument must be an object');
  }

  if (options.checkConstraints != undefined && typeof options.checkConstraints !== 'boolean') {
    throw new TypeError('"checkConstraints" must be a boolean');
  }

  if (options.fireTriggers != undefined && typeof options.fireTriggers !== 'boolean') {
    throw new TypeError('"fireTriggers" must be a boolean');
  }

  if (options.keepNulls != undefined && typeof options.keepNulls !== 'boolean') {
    throw new TypeError('"keepNulls" must be a boolean');
  }

  if (options.lockTable != undefined && typeof options.lockTable !== 'boolean') {
    throw new TypeError('"lockTable" must be a boolean');
  }

  if (options.checkConstraints != undefined) {
    this.bulkOptions.checkConstraints = options.checkConstraints;
  }

  if (options.fireTriggers != undefined) {
    this.bulkOptions.fireTriggers = options.fireTriggers;
  }

  if (options.keepNulls != undefined) {
    this.bulkOptions.keepNulls = options.keepNulls;
  }

  if (options.lockTable != undefined) {
    this.bulkOptions.lockTable = options.lockTable;
  }
}
```

This code is overall a bit longer, but fixes the aforementioned issues and is more straightforward (and thus overall easier to maintain). 👍

Reply:
I would completely replace the previous `this.bulkOptions` here, instead of modifying them:

```
setOptions(options) {
   const newOptions = {};
   newOptions.checkContraints = getBooleanOption('checkContraints', false);
   newOptions.fireTriggers = getBooleanOption('fireTriggers', false);
   newOptions.keepNulls = getBooleanOption('keepNulls', false);
   newOptions.lockTable = getBooleanOption('lockTable', false);
   this.bulkOptions = newOptions;

   function getBooleanOption(optionName, defaultValue) {
      const optionValue = options[optionName];
      if (optionValue == undefined) {
         return defaultValue;
      }
      if (typeof optionValue !== 'boolean')  {
         throw new TypeError(´"${optionName}" must be a boolean');
      }
      return optionValue;
   }
}
```

Then we can initialize `this.bulkOptions` within the class constructor with:

```
   this.setOptions({});
```

Reply:
Thank you for detailing it :)